tachyon/crypto/sumcheck/multilinear/data_structures.h

```diff
@@ -0,0 +1,167 @@
+#ifndef TACHYON_CRYPTO_SUMCHECK_MULTILINEAR_DATA_STRUCTURES_H_
+#define TACHYON_CRYPTO_SUMCHECK_MULTILINEAR_DATA_STRUCTURES_H_
+
+#include <algorithm>
+#include <memory>
+#include <numeric>
+#include <vector>
+
+#include "absl/container/node_hash_map.h"
+
+#include "tachyon/base/buffer/copyable.h"
+#include "tachyon/base/containers/contains.h"
+#include "tachyon/base/logging.h"
+#include "tachyon/math/polynomials/multivariate/multilinear_extension.h"
+
+namespace tachyon {
+
+namespace crypto::sumcheck {
+
+// This data structure will be used as the verifier key.
+struct TACHYON_EXPORT PolynomialInfo {
+  size_t max_multiplicands;
+  size_t num_variables;
+
+  bool operator==(const PolynomialInfo& other) const {
+    return max_multiplicands == other.max_multiplicands &&
+           num_variables == other.num_variables;
+  }
+  bool operator!=(const PolynomialInfo& other) const {
+    return !operator==(other);
+  }
+  // TODO(ashjeong): For better unit testing
+  // PolynomialInfo Random() {}
+};
+
+template <typename Evaluations>
+class ListOfProductsOfPolynomials;
+
+template <typename Evaluations>
+class ListOfProductsOfPolynomials<math::MultilinearExtension<Evaluations>> {
+ public:
+  using F = typename math::MultilinearExtension<Evaluations>::Field;
+
+  struct Term {
+    F coefficient;
+    std::vector<size_t> indexes;
+  };
+
+  ListOfProductsOfPolynomials() = default;
+  explicit ListOfProductsOfPolynomials(size_t num_variables)
+      : num_variables_(num_variables) {}
+  ListOfProductsOfPolynomials(ListOfProductsOfPolynomials& other) = default;
+  ListOfProductsOfPolynomials& operator=(ListOfProductsOfPolynomials& other) =
+      default;
+
+  size_t max_evaluations() { return max_evaluations_; }
+  size_t num_variables() { return num_variables_; }
+  std::vector<Term> terms() { return terms_; }
+  std::vector<std::shared_ptr<math::MultilinearExtension<Evaluations>>>
+  flattened_ml_evaluations() {
+    return flattened_ml_evaluations_;
+  }
+
+  // Adds a Term to the list
+  void AddTerm(const F& coefficient,
+               const std::vector<
+                   std::shared_ptr<math::MultilinearExtension<Evaluations>>>&
```

Comment:
Throughout the codes, we rarely uses `std::shared_ptr` so I am kind of curious how `std::shared_ptr` gets used after sumcheck!

Reply:
Looking at all the other code involved in the arkworks sumcheck repository, I don't really think the shared_ptrs are used again 🤔. On the prover side of the protocol, `flattened_ml_extensions_` is deep copied, but is instead used as a vector of `MultilinearExtensions` and not of shared_ptrs. I tried showing a better general view of what goes on in the protocol more concretely here, though I still don't understand the Prover/Verifier side fully yet. Hopefully the diagram helps a bit to understand the structure of the protocol.
![image](https://github.com/ashjeong/tachyon/assets/84874402/d15c8bb3-e4ae-4498-98c4-2fe0f4331fdc)